core/invoker/src/main/scala/org/apache/openwhisk/core/containerpool/ContainerPool.scala

```diff
@@ -340,6 +418,45 @@ class ContainerPool(childFactory: ActorRefFactory => ActorRef,
     newContainer ! Start(exec, memoryLimit)
   }
 
+  /** Create a new prewarm container if currentCount doesn't reach maxCount */
+  def prewarmContainerIfpossible(kind: String, memoryLimit: ByteSize, count: Int): Unit = {
+    prewarmConfig
+      .filter { config =>
+        kind == config.exec.kind && memoryLimit == config.memoryLimit
+      }
+      .foreach { config =>
+        val currentCount = prewarmedPool.count {
+          case (_, PreWarmedData(_, `kind`, `memoryLimit`, _)) => true //done starting
+          case _                                               => false //started but not finished starting
+        }
+        val startingCount = prewarmStartingPool.count(p => p._2._1 == kind && p._2._2 == memoryLimit)
+        val containerCount = currentCount + startingCount
+        if (containerCount < config.maxCount) {
+          val createNumber = if (config.maxCount - containerCount > count) count else config.maxCount - containerCount
+          logging.info(this, s"add ${createNumber} [kind: ${kind} memory: ${memoryLimit}] prewarmed containers")
+          1 to createNumber foreach { _ =>
+            prewarmContainer(config.exec, config.memoryLimit)
+          }
+        }
+      }
+  }
+
+  /** statistics the cold start */
+  def incrementColdStartCount(kind: String, memoryLimit: ByteSize): Unit = {
+    prewarmConfig
```

Comment:
I would clarify that this is only for cold start of prewarm configs, e.g. not blackbox or other configs. Maybe just a comment is fine, but this is very important.

Reply:
Updated accordingly